{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# nearest gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "\n",
    "\n",
    "CONFIG = Path(\"/data/ruderferlab/projects/ctcf/projects/cbs-db3/cbs-db3-annotate/config/config.yaml\")\n",
    "ENCODE_NEAREST = \"GRCh38-Closest-Genes-PC.gz\"\n",
    "# GENCODE_MAPPING = \"gencode.v47.basic.annotation.parsed.pc.longest_tx.tsv\"\n",
    "\n",
    "\n",
    "def parse_config(config_path: str) -> dict:\n",
    "    \"Parses provided YAML configuration file as a dictionary.\"\n",
    "    with open(config_path, \"r\", encoding=\"utf-8\") as file:\n",
    "        config = yaml.safe_load(file)\n",
    "    return config\n",
    "\n",
    "def parse_nearest_gene(filepath: str) -> pd.DataFrame:\n",
    "     \"\"\"d\"\"\"\n",
    "     return pd.read_csv(\n",
    "        filepath,\n",
    "        sep=\"\\t\",\n",
    "        header=None,\n",
    "        usecols=[4, 9, 12],\n",
    "        names=[\"ccre_accession\", \"tx_id\", \"gene_id\"],\n",
    "        dtype={\"ccre_accession\": str, \"tx_id\": str, \"gene_id\": str},\n",
    "        engine=\"c\",\n",
    "        compression=\"gzip\",\n",
    "        )\n",
    "\n",
    "# def parse_gencode(filepath: str) -> pd.DataFrame:\n",
    "#     \"\"\"d\"\"\"\n",
    "#     return pd.read_csv(\n",
    "#         filepath,\n",
    "#         sep=\"\\t\",\n",
    "#         header=None,\n",
    "#         names=[\"gene_name\", \"gene_id\", \"tx_id\"],\n",
    "#         dtype={\"gene_name\": str, \"gene_id\": str, \"tx_id\": str},\n",
    "#         engine=\"c\",\n",
    "#         )\n",
    "    \n",
    "###\n",
    "# Main\n",
    "###\n",
    "\n",
    "# Parse config\n",
    "try:\n",
    "    config_file = parse_config(CONFIG.resolve())\n",
    "except FileNotFoundError:\n",
    "    print(\"Error: config not found.\")\n",
    "    \n",
    "# Register inputs from resource  directory\n",
    "ENCODE_DIR = Path(config_file[\"RESOURCES\"][\"ENCODE\"][\"SCREEN\"][\"V4\"])\n",
    "# GENCODE_DIR = Path(config_file[\"RESOURCES\"][\"GENCODE\"][\"V47\"])\n",
    "\n",
    "# Register input files\n",
    "encode_nearest_gene = ENCODE_DIR / ENCODE_NEAREST\n",
    "# gencode_mapping = \"/data/ruderferlab/projects/ctcf/projects/cbs-db3/cbs-db3-annotate/resources/data/gencode/v47/gencode.v47.basic.annotation.parsed.pc.longest_tx.tsv\"\n",
    "\n",
    "# Parse input files\n",
    "nearest_gene = parse_nearest_gene(encode_nearest_gene)\n",
    "# gencode_mapping = parse_gencode(gencode_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E2776516</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E2776517</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E3951272</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E3951273</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EH38E3951274</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651782</th>\n",
       "      <td>EH38E3951270</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651783</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651784</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651785</th>\n",
       "      <td>EH38E4535243</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651786</th>\n",
       "      <td>EH38E4535244</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2651787 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ccre_accession                     tx_id                   gene_id\n",
       "0         EH38E2776516         ENST00000641515.2         ENSG00000186092.7\n",
       "1         EH38E2776517         ENST00000641515.2         ENSG00000186092.7\n",
       "2         EH38E3951272         ENST00000641515.2         ENSG00000186092.7\n",
       "3         EH38E3951273         ENST00000641515.2         ENSG00000186092.7\n",
       "4         EH38E3951274         ENST00000641515.2         ENSG00000186092.7\n",
       "...                ...                       ...                       ...\n",
       "2651782   EH38E3951270         ENST00000361963.3         ENSG00000172288.7\n",
       "2651783   EH38E4535242         ENST00000306609.4         ENSG00000172288.7\n",
       "2651784   EH38E4535242         ENST00000361963.3         ENSG00000172288.7\n",
       "2651785   EH38E4535243  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "2651786   EH38E4535244  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "\n",
       "[2651787 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###\n",
    "# For now get 1:1 mapping by dropping duplicates (development)\n",
    "###\n",
    "\n",
    "nearest_gene = nearest_gene.drop_duplicates(subset=[\"ccre_accession\"], keep=\"first\")\n",
    "\n",
    "# Trim trailing \".\" from gene_id and tx_id\n",
    "nearest_gene.loc[:, \"tx_id\"] = nearest_gene[\"tx_id\"].str.split(\".\").str[0]\n",
    "nearest_gene.loc[:, \"gene_id\"] = nearest_gene[\"gene_id\"].str.split(\".\").str[0]\n",
    "\n",
    "# Write out\n",
    "OUTDIR = \"/data/ruderferlab/projects/ctcf/projects/cbs-db3/cbs-db3-annotate/results/annotations/encode/screen/v4\"\n",
    "OUT_FILE = \"closest_pc_gene.tsv.anno.gz\"\n",
    "nearest_gene.to_csv(\n",
    "    os.path.join(OUTDIR, OUT_FILE),\n",
    "    sep=\"\\t\",\n",
    "    index=False,\n",
    "    header=True,\n",
    "    compression=\"gzip\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E2776516</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E2776517</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E3951272</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E3951273</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EH38E3951274</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651779</th>\n",
       "      <td>EH38E3951269</td>\n",
       "      <td>ENST00000306609</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651781</th>\n",
       "      <td>EH38E3951270</td>\n",
       "      <td>ENST00000306609</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651783</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000306609</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651785</th>\n",
       "      <td>EH38E4535243</td>\n",
       "      <td>ENST00000286448</td>\n",
       "      <td>ENSG00000124333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651786</th>\n",
       "      <td>EH38E4535244</td>\n",
       "      <td>ENST00000286448</td>\n",
       "      <td>ENSG00000124333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2348854 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ccre_accession            tx_id          gene_id\n",
       "0         EH38E2776516  ENST00000641515  ENSG00000186092\n",
       "1         EH38E2776517  ENST00000641515  ENSG00000186092\n",
       "2         EH38E3951272  ENST00000641515  ENSG00000186092\n",
       "3         EH38E3951273  ENST00000641515  ENSG00000186092\n",
       "4         EH38E3951274  ENST00000641515  ENSG00000186092\n",
       "...                ...              ...              ...\n",
       "2651779   EH38E3951269  ENST00000306609  ENSG00000172288\n",
       "2651781   EH38E3951270  ENST00000306609  ENSG00000172288\n",
       "2651783   EH38E4535242  ENST00000306609  ENSG00000172288\n",
       "2651785   EH38E4535243  ENST00000286448  ENSG00000124333\n",
       "2651786   EH38E4535244  ENST00000286448  ENSG00000124333\n",
       "\n",
       "[2348854 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tx_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR4F16</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>ENST00000327044.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20087</th>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>ENST00000361335.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20088</th>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>ENST00000361381.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20089</th>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>ENST00000361567.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20090</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20091</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20092 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name             gene_id              tx_id\n",
       "0         OR4F5   ENSG00000186092.7  ENST00000641515.2\n",
       "1        OR4F29   ENSG00000284733.2  ENST00000426406.4\n",
       "2        OR4F16   ENSG00000284662.2  ENST00000332831.5\n",
       "3        SAMD11  ENSG00000187634.13  ENST00000616016.5\n",
       "4         NOC2L  ENSG00000188976.11  ENST00000327044.7\n",
       "...         ...                 ...                ...\n",
       "20087   MT-ND4L   ENSG00000212907.2  ENST00000361335.1\n",
       "20088    MT-ND4   ENSG00000198886.2  ENST00000361381.2\n",
       "20089    MT-ND5   ENSG00000198786.2  ENST00000361567.2\n",
       "20090    MT-ND6   ENSG00000198695.2  ENST00000361681.2\n",
       "20091    MT-CYB   ENSG00000198727.2  ENST00000361789.2\n",
       "\n",
       "[20092 rows x 3 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ccre_accession\n",
       "EH38E3297036    4\n",
       "EH38E1951203    4\n",
       "EH38E2571346    4\n",
       "EH38E3024082    3\n",
       "EH38E2761830    3\n",
       "               ..\n",
       "EH38E3172994    1\n",
       "EH38E3172993    1\n",
       "EH38E3172992    1\n",
       "EH38E1807374    1\n",
       "EH38E4535244    1\n",
       "Name: count, Length: 2348854, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene.drop_duplicates()['ccre_accession'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1144859</th>\n",
       "      <td>EH38E3297036</td>\n",
       "      <td>ENST00000444486.7</td>\n",
       "      <td>ENSG00000213999.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1144860</th>\n",
       "      <td>EH38E3297036</td>\n",
       "      <td>ENST00000514819.7</td>\n",
       "      <td>ENSG00000064489.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1144861</th>\n",
       "      <td>EH38E3297036</td>\n",
       "      <td>ENST00000585679.1</td>\n",
       "      <td>ENSG00000254901.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1144862</th>\n",
       "      <td>EH38E3297036</td>\n",
       "      <td>ENST00000462790.8</td>\n",
       "      <td>ENSG00000254901.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1144863</th>\n",
       "      <td>EH38E3297036</td>\n",
       "      <td>ENST00000456252.7</td>\n",
       "      <td>ENSG00000064490.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1144864</th>\n",
       "      <td>EH38E3297036</td>\n",
       "      <td>ENST00000303088.9</td>\n",
       "      <td>ENSG00000064490.14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ccre_accession              tx_id             gene_id\n",
       "1144859   EH38E3297036  ENST00000444486.7  ENSG00000213999.17\n",
       "1144860   EH38E3297036  ENST00000514819.7  ENSG00000064489.23\n",
       "1144861   EH38E3297036  ENST00000585679.1   ENSG00000254901.8\n",
       "1144862   EH38E3297036  ENST00000462790.8   ENSG00000254901.8\n",
       "1144863   EH38E3297036  ENST00000456252.7  ENSG00000064490.14\n",
       "1144864   EH38E3297036  ENST00000303088.9  ENSG00000064490.14"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene[nearest_gene['ccre_accession']==\"EH38E3297036\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E2776516</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E2776517</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E3951272</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E3951273</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EH38E3951274</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651779</th>\n",
       "      <td>EH38E3951269</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651781</th>\n",
       "      <td>EH38E3951270</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651783</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651785</th>\n",
       "      <td>EH38E4535243</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651786</th>\n",
       "      <td>EH38E4535244</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2348854 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ccre_accession                     tx_id                   gene_id\n",
       "0         EH38E2776516         ENST00000641515.2         ENSG00000186092.7\n",
       "1         EH38E2776517         ENST00000641515.2         ENSG00000186092.7\n",
       "2         EH38E3951272         ENST00000641515.2         ENSG00000186092.7\n",
       "3         EH38E3951273         ENST00000641515.2         ENSG00000186092.7\n",
       "4         EH38E3951274         ENST00000641515.2         ENSG00000186092.7\n",
       "...                ...                       ...                       ...\n",
       "2651779   EH38E3951269         ENST00000306609.4         ENSG00000172288.7\n",
       "2651781   EH38E3951270         ENST00000306609.4         ENSG00000172288.7\n",
       "2651783   EH38E4535242         ENST00000306609.4         ENSG00000172288.7\n",
       "2651785   EH38E4535243  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "2651786   EH38E4535244  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "\n",
       "[2348854 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene.drop_duplicates(subset=[\"ccre_accession\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E2776516</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E2776517</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E3951272</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E3951273</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EH38E3951274</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651782</th>\n",
       "      <td>EH38E3951270</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651783</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651784</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651785</th>\n",
       "      <td>EH38E4535243</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651786</th>\n",
       "      <td>EH38E4535244</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2651787 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ccre_accession                     tx_id                   gene_id\n",
       "0         EH38E2776516         ENST00000641515.2         ENSG00000186092.7\n",
       "1         EH38E2776517         ENST00000641515.2         ENSG00000186092.7\n",
       "2         EH38E3951272         ENST00000641515.2         ENSG00000186092.7\n",
       "3         EH38E3951273         ENST00000641515.2         ENSG00000186092.7\n",
       "4         EH38E3951274         ENST00000641515.2         ENSG00000186092.7\n",
       "...                ...                       ...                       ...\n",
       "2651782   EH38E3951270         ENST00000361963.3         ENSG00000172288.7\n",
       "2651783   EH38E4535242         ENST00000306609.4         ENSG00000172288.7\n",
       "2651784   EH38E4535242         ENST00000361963.3         ENSG00000172288.7\n",
       "2651785   EH38E4535243  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "2651786   EH38E4535244  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "\n",
       "[2651787 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = nearest_gene.merge(\n",
    "    gencode_mapping,\n",
    "    how=\"left\",\n",
    "    on=['gene_id'],\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id_x</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tx_id_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E2776516</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E2776517</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E3951272</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E3951273</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EH38E3951274</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651782</th>\n",
       "      <td>EH38E3951270</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651783</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651784</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651785</th>\n",
       "      <td>EH38E4535243</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651786</th>\n",
       "      <td>EH38E4535244</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2651787 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ccre_accession                   tx_id_x                   gene_id  \\\n",
       "0         EH38E2776516         ENST00000641515.2         ENSG00000186092.7   \n",
       "1         EH38E2776517         ENST00000641515.2         ENSG00000186092.7   \n",
       "2         EH38E3951272         ENST00000641515.2         ENSG00000186092.7   \n",
       "3         EH38E3951273         ENST00000641515.2         ENSG00000186092.7   \n",
       "4         EH38E3951274         ENST00000641515.2         ENSG00000186092.7   \n",
       "...                ...                       ...                       ...   \n",
       "2651782   EH38E3951270         ENST00000361963.3         ENSG00000172288.7   \n",
       "2651783   EH38E4535242         ENST00000306609.4         ENSG00000172288.7   \n",
       "2651784   EH38E4535242         ENST00000361963.3         ENSG00000172288.7   \n",
       "2651785   EH38E4535243  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y   \n",
       "2651786   EH38E4535244  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y   \n",
       "\n",
       "        gene_name            tx_id_y  \n",
       "0           OR4F5  ENST00000641515.2  \n",
       "1           OR4F5  ENST00000641515.2  \n",
       "2           OR4F5  ENST00000641515.2  \n",
       "3           OR4F5  ENST00000641515.2  \n",
       "4           OR4F5  ENST00000641515.2  \n",
       "...           ...                ...  \n",
       "2651782       NaN                NaN  \n",
       "2651783       NaN                NaN  \n",
       "2651784       NaN                NaN  \n",
       "2651785       NaN                NaN  \n",
       "2651786       NaN                NaN  \n",
       "\n",
       "[2651787 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2348854"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene['ccre_accession'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E2776516</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E2776517</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E3951272</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E3951273</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EH38E3951274</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ccre_accession              tx_id            gene_id\n",
       "0   EH38E2776516  ENST00000641515.2  ENSG00000186092.7\n",
       "1   EH38E2776517  ENST00000641515.2  ENSG00000186092.7\n",
       "2   EH38E3951272  ENST00000641515.2  ENSG00000186092.7\n",
       "3   EH38E3951273  ENST00000641515.2  ENSG00000186092.7\n",
       "4   EH38E3951274  ENST00000641515.2  ENSG00000186092.7"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ccre_accession\n",
       "EH38E3975884    47\n",
       "EH38E2320864    45\n",
       "EH38E2309980    42\n",
       "EH38E3818239    40\n",
       "EH38E3539202    37\n",
       "                ..\n",
       "EH38E3190762     1\n",
       "EH38E4148422     1\n",
       "EH38E3190760     1\n",
       "EH38E3190759     1\n",
       "EH38E4535244     1\n",
       "Name: count, Length: 2348854, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene['ccre_accession'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>114671</th>\n",
       "      <td>EH38E3975884</td>\n",
       "      <td>ENST00000685628.1</td>\n",
       "      <td>ENSG00000121940.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114672</th>\n",
       "      <td>EH38E3975884</td>\n",
       "      <td>ENST00000676454.1</td>\n",
       "      <td>ENSG00000121940.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114673</th>\n",
       "      <td>EH38E3975884</td>\n",
       "      <td>ENST00000675790.1</td>\n",
       "      <td>ENSG00000121940.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114674</th>\n",
       "      <td>EH38E3975884</td>\n",
       "      <td>ENST00000674849.1</td>\n",
       "      <td>ENSG00000121940.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114675</th>\n",
       "      <td>EH38E3975884</td>\n",
       "      <td>ENST00000675956.1</td>\n",
       "      <td>ENSG00000121940.17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ccre_accession              tx_id             gene_id\n",
       "114671   EH38E3975884  ENST00000685628.1  ENSG00000121940.17\n",
       "114672   EH38E3975884  ENST00000676454.1  ENSG00000121940.17\n",
       "114673   EH38E3975884  ENST00000675790.1  ENSG00000121940.17\n",
       "114674   EH38E3975884  ENST00000674849.1  ENSG00000121940.17\n",
       "114675   EH38E3975884  ENST00000675956.1  ENSG00000121940.17"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene[nearest_gene['ccre_accession'] == 'EH38E3975884'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tx_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651782</th>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651783</th>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651784</th>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651785</th>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651786</th>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2651787 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_id                     tx_id\n",
       "0               ENSG00000186092.7         ENST00000641515.2\n",
       "1               ENSG00000186092.7         ENST00000641515.2\n",
       "2               ENSG00000186092.7         ENST00000641515.2\n",
       "3               ENSG00000186092.7         ENST00000641515.2\n",
       "4               ENSG00000186092.7         ENST00000641515.2\n",
       "...                           ...                       ...\n",
       "2651782         ENSG00000172288.7         ENST00000361963.3\n",
       "2651783         ENSG00000172288.7         ENST00000306609.4\n",
       "2651784         ENSG00000172288.7         ENST00000361963.3\n",
       "2651785  ENSG00000124333.16_PAR_Y  ENST00000286448.12_PAR_Y\n",
       "2651786  ENSG00000124333.16_PAR_Y  ENST00000286448.12_PAR_Y\n",
       "\n",
       "[2651787 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene[['gene_id','tx_id']].merge(gencode_mapping[['gene_id','tx_id']], on=['gene_id','tx_id'], how='left').dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name\n",
       "NaN                1606215\n",
       "TENM3                 1688\n",
       "TENM4                 1430\n",
       "SOX9                  1345\n",
       "UQCRFS1               1337\n",
       "                    ...   \n",
       "PSENEN                   1\n",
       "U2AF1L4                  1\n",
       "ENSG00000267120          1\n",
       "IGFLR1                   1\n",
       "ENSG00000285505          1\n",
       "Name: count, Length: 17300, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene.merge(\n",
    "    gencode_mapping,\n",
    "    how=\"left\",\n",
    "    on=['gene_id', 'tx_id'])['gene_name'].value_counts(dropna=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tx_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR4F16</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>ENST00000327044.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name             gene_id              tx_id\n",
       "0     OR4F5   ENSG00000186092.7  ENST00000641515.2\n",
       "1    OR4F29   ENSG00000284733.2  ENST00000426406.4\n",
       "2    OR4F16   ENSG00000284662.2  ENST00000332831.5\n",
       "3    SAMD11  ENSG00000187634.13  ENST00000616016.5\n",
       "4     NOC2L  ENSG00000188976.11  ENST00000327044.7"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>72702</th>\n",
       "      <td>EH38E0001314</td>\n",
       "      <td>ENST00000371226.8</td>\n",
       "      <td>ENSG00000162600.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170113</th>\n",
       "      <td>EH38E0003674</td>\n",
       "      <td>ENST00000367462.5</td>\n",
       "      <td>ENSG00000162670.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170281</th>\n",
       "      <td>EH38E0003686</td>\n",
       "      <td>ENST00000367462.5</td>\n",
       "      <td>ENSG00000162670.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197068</th>\n",
       "      <td>EH38E0004320</td>\n",
       "      <td>ENST00000259154.9</td>\n",
       "      <td>ENSG00000136636.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240234</th>\n",
       "      <td>EH38E0005245</td>\n",
       "      <td>ENST00000298375.12</td>\n",
       "      <td>ENSG00000165568.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651773</th>\n",
       "      <td>EH38E4535241</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651783</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651784</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651785</th>\n",
       "      <td>EH38E4535243</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651786</th>\n",
       "      <td>EH38E4535244</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2651787 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ccre_accession                     tx_id                   gene_id\n",
       "72702     EH38E0001314         ENST00000371226.8        ENSG00000162600.12\n",
       "170113    EH38E0003674         ENST00000367462.5        ENSG00000162670.11\n",
       "170281    EH38E0003686         ENST00000367462.5        ENSG00000162670.11\n",
       "197068    EH38E0004320         ENST00000259154.9        ENSG00000136636.13\n",
       "240234    EH38E0005245        ENST00000298375.12        ENSG00000165568.18\n",
       "...                ...                       ...                       ...\n",
       "2651773   EH38E4535241         ENST00000306609.4         ENSG00000172288.7\n",
       "2651783   EH38E4535242         ENST00000306609.4         ENSG00000172288.7\n",
       "2651784   EH38E4535242         ENST00000361963.3         ENSG00000172288.7\n",
       "2651785   EH38E4535243  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "2651786   EH38E4535244  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "\n",
       "[2651787 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene.sort_values(\n",
    "    by=[\"ccre_accession\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Register input resource  directory\n",
    "ip_dir = Path(config_file[\"FILEPATHS\"][\"RESOURCE_DIR\"]).resolve()\n",
    "\n",
    "# Register input resource file\n",
    "ip_file = ip_dir / IP_DATA\n",
    "\n",
    "# Parse input resource file\n",
    "nearest_gene = parse_nearest_gene(ip_file)\n",
    "\n",
    "# Create 1 to 1 mapping of ccre_accession to gene/tx\n",
    "\n",
    "# Sort\n",
    "nearest_gene = nearest_gene.sort_values(by=[\"ccre_accession\", \"tx_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccre_accession</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>72702</th>\n",
       "      <td>EH38E0001314</td>\n",
       "      <td>ENST00000371226.8</td>\n",
       "      <td>ENSG00000162600.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170113</th>\n",
       "      <td>EH38E0003674</td>\n",
       "      <td>ENST00000367462.5</td>\n",
       "      <td>ENSG00000162670.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170281</th>\n",
       "      <td>EH38E0003686</td>\n",
       "      <td>ENST00000367462.5</td>\n",
       "      <td>ENSG00000162670.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197068</th>\n",
       "      <td>EH38E0004320</td>\n",
       "      <td>ENST00000259154.9</td>\n",
       "      <td>ENSG00000136636.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240234</th>\n",
       "      <td>EH38E0005245</td>\n",
       "      <td>ENST00000298375.12</td>\n",
       "      <td>ENSG00000165568.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651774</th>\n",
       "      <td>EH38E4535241</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651783</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000306609.4</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651784</th>\n",
       "      <td>EH38E4535242</td>\n",
       "      <td>ENST00000361963.3</td>\n",
       "      <td>ENSG00000172288.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651785</th>\n",
       "      <td>EH38E4535243</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651786</th>\n",
       "      <td>EH38E4535244</td>\n",
       "      <td>ENST00000286448.12_PAR_Y</td>\n",
       "      <td>ENSG00000124333.16_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2651787 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ccre_accession                     tx_id                   gene_id\n",
       "72702     EH38E0001314         ENST00000371226.8        ENSG00000162600.12\n",
       "170113    EH38E0003674         ENST00000367462.5        ENSG00000162670.11\n",
       "170281    EH38E0003686         ENST00000367462.5        ENSG00000162670.11\n",
       "197068    EH38E0004320         ENST00000259154.9        ENSG00000136636.13\n",
       "240234    EH38E0005245        ENST00000298375.12        ENSG00000165568.18\n",
       "...                ...                       ...                       ...\n",
       "2651774   EH38E4535241         ENST00000361963.3         ENSG00000172288.7\n",
       "2651783   EH38E4535242         ENST00000306609.4         ENSG00000172288.7\n",
       "2651784   EH38E4535242         ENST00000361963.3         ENSG00000172288.7\n",
       "2651785   EH38E4535243  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "2651786   EH38E4535244  ENST00000286448.12_PAR_Y  ENSG00000124333.16_PAR_Y\n",
       "\n",
       "[2651787 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nearest_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3D chromatin gene target"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3D Chromatin\n",
    "| cCRE ID | Gene ID | Common Gene Name | Gene Type | Assay Type | Experiment ID | Biosample | Score | P-value |\n",
    "|---------|---------|-----------------|-----------|------------|---------------|-----------|--------|----------|\n",
    "| | | | | | | | | |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENE_TARGETS = \"/panfs/accrepfs.vampire/data/ruderferlab/projects/ctcf/resources/data/encode/screen/v4/V4-hg38.Gene-Links.3D-Chromatin.txt.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read gene targets\n",
    "gene_targets = pd.read_csv(\n",
    "    GENE_TARGETS,\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    # usecols=[0, 1, 2, 3],\n",
    "    # names=[\"ccre_accession\", \"gene_id\", \"tx_id\", \"gene_name\"],\n",
    "    # dtype={\"ccre_accession\": str, \"gene_id\": str, \"tx_id\": str, \"gene_name\": str},\n",
    "    engine=\"c\",\n",
    "    compression=\"gzip\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E0080197</td>\n",
       "      <td>ENSG00000121769</td>\n",
       "      <td>FABP3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>1.31</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E0080197</td>\n",
       "      <td>ENSG00000134644</td>\n",
       "      <td>PUM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>10.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E0082273</td>\n",
       "      <td>ENSG00000171812</td>\n",
       "      <td>COL8A2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>15.68</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E0096620</td>\n",
       "      <td>ENSG00000118454</td>\n",
       "      <td>ANKRD13C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>1.61</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EH38E0096620</td>\n",
       "      <td>ENSG00000197568</td>\n",
       "      <td>HHLA3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>1.61</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0                1          2               3               4  \\\n",
       "0  EH38E0080197  ENSG00000121769     FABP3   protein_coding  RNAPII-ChIAPET   \n",
       "1  EH38E0080197  ENSG00000134644      PUM1   protein_coding  RNAPII-ChIAPET   \n",
       "2  EH38E0082273  ENSG00000171812    COL8A2   protein_coding  RNAPII-ChIAPET   \n",
       "3  EH38E0096620  ENSG00000118454  ANKRD13C   protein_coding  RNAPII-ChIAPET   \n",
       "4  EH38E0096620  ENSG00000197568     HHLA3           lncRNA  RNAPII-ChIAPET   \n",
       "\n",
       "             5   6      7   8  \n",
       "0  ENCSR782EKZ  H1   1.31 NaN  \n",
       "1  ENCSR782EKZ  H1  10.39 NaN  \n",
       "2  ENCSR782EKZ  H1  15.68 NaN  \n",
       "3  ENCSR782EKZ  H1   1.61 NaN  \n",
       "4  ENCSR782EKZ  H1   1.61 NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_targets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subet to protein coding\n",
    "gene_targets = gene_targets[gene_targets[3] == \"protein_coding\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E0048077</td>\n",
       "      <td>ENSG00000198755</td>\n",
       "      <td>HEK293</td>\n",
       "      <td>6.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E0048077</td>\n",
       "      <td>ENSG00000198755</td>\n",
       "      <td>HEK293T</td>\n",
       "      <td>3.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E0048077</td>\n",
       "      <td>ENSG00000198755</td>\n",
       "      <td>WTC11_genetically_modified_(insertion)_using_C...</td>\n",
       "      <td>2.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E0055676</td>\n",
       "      <td>ENSG00000183779</td>\n",
       "      <td>MCF-7</td>\n",
       "      <td>8.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EH38E0070245</td>\n",
       "      <td>ENSG00000162496</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>9.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3813857</th>\n",
       "      <td>EH38E4535008</td>\n",
       "      <td>ENSG00000288049</td>\n",
       "      <td>naive_thymus-derived_CD4-positive,_alpha-beta_...</td>\n",
       "      <td>14.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3813858</th>\n",
       "      <td>EH38E4535008</td>\n",
       "      <td>ENSG00000288049</td>\n",
       "      <td>stimulated_activated_naive_CD4-positive,_alpha...</td>\n",
       "      <td>8.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3813859</th>\n",
       "      <td>EH38E4535015</td>\n",
       "      <td>ENSG00000131002</td>\n",
       "      <td>OCI-LY7</td>\n",
       "      <td>2.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3813860</th>\n",
       "      <td>EH38E4535017</td>\n",
       "      <td>ENSG00000012817</td>\n",
       "      <td>WTC11_genetically_modified_(insertion)_using_C...</td>\n",
       "      <td>2.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3813861</th>\n",
       "      <td>EH38E4535017</td>\n",
       "      <td>ENSG00000288049</td>\n",
       "      <td>WTC11_genetically_modified_(insertion)_using_C...</td>\n",
       "      <td>2.72</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3813862 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0                1  \\\n",
       "0        EH38E0048077  ENSG00000198755   \n",
       "1        EH38E0048077  ENSG00000198755   \n",
       "2        EH38E0048077  ENSG00000198755   \n",
       "3        EH38E0055676  ENSG00000183779   \n",
       "4        EH38E0070245  ENSG00000162496   \n",
       "...               ...              ...   \n",
       "3813857  EH38E4535008  ENSG00000288049   \n",
       "3813858  EH38E4535008  ENSG00000288049   \n",
       "3813859  EH38E4535015  ENSG00000131002   \n",
       "3813860  EH38E4535017  ENSG00000012817   \n",
       "3813861  EH38E4535017  ENSG00000288049   \n",
       "\n",
       "                                                         6      7  \n",
       "0                                                   HEK293   6.38  \n",
       "1                                                  HEK293T   3.03  \n",
       "2        WTC11_genetically_modified_(insertion)_using_C...   2.64  \n",
       "3                                                    MCF-7   8.28  \n",
       "4                                                   HCT116   9.59  \n",
       "...                                                    ...    ...  \n",
       "3813857  naive_thymus-derived_CD4-positive,_alpha-beta_...  14.44  \n",
       "3813858  stimulated_activated_naive_CD4-positive,_alpha...   8.04  \n",
       "3813859                                            OCI-LY7   2.35  \n",
       "3813860  WTC11_genetically_modified_(insertion)_using_C...   2.72  \n",
       "3813861  WTC11_genetically_modified_(insertion)_using_C...   2.72  \n",
       "\n",
       "[3813862 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for each ccre-gene-biosample, select the highest score\n",
    "highest_scores = gene_targets.groupby(by=[0, 1, 6])[7].max().reset_index()\n",
    "highest_scores = highest_scores.rename(columns={0: \"ccre_accession\", 1: \"gene_id\", 6: \"biosample_id\", 7: \"score\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EH38E0080197</td>\n",
       "      <td>ENSG00000121769</td>\n",
       "      <td>FABP3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>1.31</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EH38E0080197</td>\n",
       "      <td>ENSG00000134644</td>\n",
       "      <td>PUM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>10.39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EH38E0082273</td>\n",
       "      <td>ENSG00000171812</td>\n",
       "      <td>COL8A2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>15.68</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EH38E0096620</td>\n",
       "      <td>ENSG00000118454</td>\n",
       "      <td>ANKRD13C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>1.61</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EH38E0114249</td>\n",
       "      <td>ENSG00000178096</td>\n",
       "      <td>BOLA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR782EKZ</td>\n",
       "      <td>H1</td>\n",
       "      <td>21.65</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4168059</th>\n",
       "      <td>EH38E1545247</td>\n",
       "      <td>ENSG00000126391</td>\n",
       "      <td>FRMD8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR499JGQ</td>\n",
       "      <td>MCF_10A</td>\n",
       "      <td>62.75</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4168060</th>\n",
       "      <td>EH38E1545247</td>\n",
       "      <td>ENSG00000142186</td>\n",
       "      <td>SCYL1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR499JGQ</td>\n",
       "      <td>MCF_10A</td>\n",
       "      <td>3.08</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4168061</th>\n",
       "      <td>EH38E1545247</td>\n",
       "      <td>ENSG00000162241</td>\n",
       "      <td>SLC25A45</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR499JGQ</td>\n",
       "      <td>MCF_10A</td>\n",
       "      <td>62.75</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4168062</th>\n",
       "      <td>EH38E1545247</td>\n",
       "      <td>ENSG00000173465</td>\n",
       "      <td>ZNRD2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR499JGQ</td>\n",
       "      <td>MCF_10A</td>\n",
       "      <td>2.53</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4168064</th>\n",
       "      <td>EH38E1545247</td>\n",
       "      <td>ENSG00000176973</td>\n",
       "      <td>FAM89B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RNAPII-ChIAPET</td>\n",
       "      <td>ENCSR499JGQ</td>\n",
       "      <td>MCF_10A</td>\n",
       "      <td>2.53</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2907565 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0                1          2               3  \\\n",
       "0        EH38E0080197  ENSG00000121769     FABP3   protein_coding   \n",
       "1        EH38E0080197  ENSG00000134644      PUM1   protein_coding   \n",
       "2        EH38E0082273  ENSG00000171812    COL8A2   protein_coding   \n",
       "3        EH38E0096620  ENSG00000118454  ANKRD13C   protein_coding   \n",
       "5        EH38E0114249  ENSG00000178096     BOLA1   protein_coding   \n",
       "...               ...              ...        ...             ...   \n",
       "4168059  EH38E1545247  ENSG00000126391     FRMD8   protein_coding   \n",
       "4168060  EH38E1545247  ENSG00000142186     SCYL1   protein_coding   \n",
       "4168061  EH38E1545247  ENSG00000162241  SLC25A45   protein_coding   \n",
       "4168062  EH38E1545247  ENSG00000173465     ZNRD2   protein_coding   \n",
       "4168064  EH38E1545247  ENSG00000176973    FAM89B   protein_coding   \n",
       "\n",
       "                      4            5        6      7   8  \n",
       "0        RNAPII-ChIAPET  ENCSR782EKZ       H1   1.31 NaN  \n",
       "1        RNAPII-ChIAPET  ENCSR782EKZ       H1  10.39 NaN  \n",
       "2        RNAPII-ChIAPET  ENCSR782EKZ       H1  15.68 NaN  \n",
       "3        RNAPII-ChIAPET  ENCSR782EKZ       H1   1.61 NaN  \n",
       "5        RNAPII-ChIAPET  ENCSR782EKZ       H1  21.65 NaN  \n",
       "...                 ...          ...      ...    ...  ..  \n",
       "4168059  RNAPII-ChIAPET  ENCSR499JGQ  MCF_10A  62.75 NaN  \n",
       "4168060  RNAPII-ChIAPET  ENCSR499JGQ  MCF_10A   3.08 NaN  \n",
       "4168061  RNAPII-ChIAPET  ENCSR499JGQ  MCF_10A  62.75 NaN  \n",
       "4168062  RNAPII-ChIAPET  ENCSR499JGQ  MCF_10A   2.53 NaN  \n",
       "4168064  RNAPII-ChIAPET  ENCSR499JGQ  MCF_10A   2.53 NaN  \n",
       "\n",
       "[2907565 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_targets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# biosamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "BIOSAMPLE_MAP = \"/data/ruderferlab/projects/ctcf/projects/cbs-db3/cbs-db3-annotate/resources/data/encode/screen/v4/biosample_map.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read biosample map\n",
    "biosample_map = pd.read_csv(\n",
    "    BIOSAMPLE_MAP,\n",
    "    sep=\"\\t\",\n",
    "    engine=\"c\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "####\n",
    "# Main\n",
    "###\n",
    "\n",
    "# Extract experiment ID\n",
    "biosample_map[\"experiment_id\"] = biosample_map[\"value\"].str.split(\"_\").str[-1]\n",
    "\n",
    "# Drop sampleType\n",
    "biosample_map.drop(columns=[\"sampleType\"], axis=1, inplace=True)\n",
    "\n",
    "# Set column labels\n",
    "labels = [\"biosample\", \"tissue\", \"life_stage\", \"display_name\", \"experiment_id\"]\n",
    "biosample_map.columns = labels\n",
    "\n",
    "# Write out\n",
    "biosample_map.to_csv(\"/data/ruderferlab/projects/ctcf/projects/cbs-db3/cbs-db3-annotate/results/tables/encode/screen/v4/biosample_map.tsv.table.gz\", sep=\"\\t\", index=False, header=True, compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>life_stage</th>\n",
       "      <th>display_name</th>\n",
       "      <th>experiment_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22Rv1_ENCDO719HIY</td>\n",
       "      <td>prostate</td>\n",
       "      <td>adult</td>\n",
       "      <td>22Rv1</td>\n",
       "      <td>ENCDO719HIY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22Rv1_treated_with_10_nM_17β-hydroxy-5α-andros...</td>\n",
       "      <td>prostate</td>\n",
       "      <td>adult</td>\n",
       "      <td>22Rv1, treated with 10 nM 17β-hydroxy-5α-andro...</td>\n",
       "      <td>ENCDO719HIY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8988T_ENCDO000ACP</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>adult</td>\n",
       "      <td>8988T</td>\n",
       "      <td>ENCDO000ACP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A172_ENCDO934VEN</td>\n",
       "      <td>brain</td>\n",
       "      <td>adult</td>\n",
       "      <td>A172</td>\n",
       "      <td>ENCDO934VEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A549_ENCDO000AAZ</td>\n",
       "      <td>lung</td>\n",
       "      <td>adult</td>\n",
       "      <td>A549</td>\n",
       "      <td>ENCDO000AAZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1883</th>\n",
       "      <td>with_squamous_cell_carcinoma_skin_epidermis_ti...</td>\n",
       "      <td>skin</td>\n",
       "      <td>adult</td>\n",
       "      <td>skin epidermis, male adult (75 years) with squ...</td>\n",
       "      <td>ENCDO317WFZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1884</th>\n",
       "      <td>with_squamous_cell_carcinoma_skin_epidermis_ti...</td>\n",
       "      <td>skin</td>\n",
       "      <td>adult</td>\n",
       "      <td>skin epidermis, male adult (78 years) with squ...</td>\n",
       "      <td>ENCDO054KII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1885</th>\n",
       "      <td>with_squamous_cell_carcinoma_skin_epidermis_ti...</td>\n",
       "      <td>skin</td>\n",
       "      <td>adult</td>\n",
       "      <td>skin epidermis, male adult (84 years) with squ...</td>\n",
       "      <td>ENCDO799UNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1886</th>\n",
       "      <td>WTC11_ENCDO882UJI</td>\n",
       "      <td>skin</td>\n",
       "      <td>adult</td>\n",
       "      <td>WTC11</td>\n",
       "      <td>ENCDO882UJI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1887</th>\n",
       "      <td>WTC11_genetically_modified_insertion_using_TAL...</td>\n",
       "      <td>skin</td>\n",
       "      <td>adult</td>\n",
       "      <td>WTC11, genetically modified (insertion) using ...</td>\n",
       "      <td>ENCDO882UJI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1888 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              biosample    tissue life_stage  \\\n",
       "0                                     22Rv1_ENCDO719HIY  prostate      adult   \n",
       "1     22Rv1_treated_with_10_nM_17β-hydroxy-5α-andros...  prostate      adult   \n",
       "2                                     8988T_ENCDO000ACP  pancreas      adult   \n",
       "3                                      A172_ENCDO934VEN     brain      adult   \n",
       "4                                      A549_ENCDO000AAZ      lung      adult   \n",
       "...                                                 ...       ...        ...   \n",
       "1883  with_squamous_cell_carcinoma_skin_epidermis_ti...      skin      adult   \n",
       "1884  with_squamous_cell_carcinoma_skin_epidermis_ti...      skin      adult   \n",
       "1885  with_squamous_cell_carcinoma_skin_epidermis_ti...      skin      adult   \n",
       "1886                                  WTC11_ENCDO882UJI      skin      adult   \n",
       "1887  WTC11_genetically_modified_insertion_using_TAL...      skin      adult   \n",
       "\n",
       "                                           display_name experiment_id  \n",
       "0                                                 22Rv1   ENCDO719HIY  \n",
       "1     22Rv1, treated with 10 nM 17β-hydroxy-5α-andro...   ENCDO719HIY  \n",
       "2                                                 8988T   ENCDO000ACP  \n",
       "3                                                  A172   ENCDO934VEN  \n",
       "4                                                  A549   ENCDO000AAZ  \n",
       "...                                                 ...           ...  \n",
       "1883  skin epidermis, male adult (75 years) with squ...   ENCDO317WFZ  \n",
       "1884  skin epidermis, male adult (78 years) with squ...   ENCDO054KII  \n",
       "1885  skin epidermis, male adult (84 years) with squ...   ENCDO799UNP  \n",
       "1886                                              WTC11   ENCDO882UJI  \n",
       "1887  WTC11, genetically modified (insertion) using ...   ENCDO882UJI  \n",
       "\n",
       "[1888 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biosample_map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gencode parse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gtfparse import read_gtf\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'tag', 'transcript_id', 'transcript_type', 'transcript_name', 'exon_number', 'exon_id', 'hgnc_id', 'havana_gene', 'transcript_support_level', 'ont', 'havana_transcript', 'protein_id', 'ccdsid', 'artif_dupl']\n"
     ]
    }
   ],
   "source": [
    "gencode_raw = \"/panfs/accrepfs.vampire/data/ruderferlab/projects/ctcf/resources/data/gencode/v47/gencode.v47.basic.annotation.gtf\"\n",
    "gencode = read_gtf(gencode_raw).to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ont</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11121</td>\n",
       "      <td>24894</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11426</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11426</td>\n",
       "      <td>11671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004248702.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004248735.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  start    end  score strand  frame  \\\n",
       "0    chr1  HAVANA        gene  11121  24894    NaN      +      0   \n",
       "1    chr1  HAVANA  transcript  11426  14409    NaN      +      0   \n",
       "2    chr1  HAVANA        exon  11426  11671    NaN      +      0   \n",
       "3    chr1  HAVANA        exon  12010  12227    NaN      +      0   \n",
       "4    chr1  HAVANA        exon  12613  12721    NaN      +      0   \n",
       "\n",
       "             gene_id gene_type  ... exon_number            exon_id hgnc_id  \\\n",
       "0  ENSG00000290825.2    lncRNA  ...                                          \n",
       "1  ENSG00000290825.2    lncRNA  ...                                          \n",
       "2  ENSG00000290825.2    lncRNA  ...           1  ENSE00004248702.1           \n",
       "3  ENSG00000290825.2    lncRNA  ...           2  ENSE00004248735.1           \n",
       "4  ENSG00000290825.2    lncRNA  ...           3  ENSE00003582793.1           \n",
       "\n",
       "  havana_gene transcript_support_level ont havana_transcript protein_id  \\\n",
       "0                                                                         \n",
       "1                                                                         \n",
       "2                                                                         \n",
       "3                                                                         \n",
       "4                                                                         \n",
       "\n",
       "  ccdsid artif_dupl  \n",
       "0                    \n",
       "1                    \n",
       "2                    \n",
       "3                    \n",
       "4                    \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecode_tx = gencode[(gencode['feature'] == 'transcript') & (gencode['gene_type'] == 'protein_coding')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecode_sub = genecode_tx[[\"transcript_id\",  \"gene_id\", \"gene_name\", \"gene_type\", \"seqname\", \"start\", \"end\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_231787/217293794.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genecode_sub['length'] = genecode_sub['end'] - genecode_sub['start']\n"
     ]
    }
   ],
   "source": [
    "genecode_sub['length'] = genecode_sub['end'] - genecode_sub['start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>6166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>923923</td>\n",
       "      <td>944574</td>\n",
       "      <td>20651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>ENST00000618323.5</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>923923</td>\n",
       "      <td>944574</td>\n",
       "      <td>20651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181</th>\n",
       "      <td>ENST00000342066.8</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>925731</td>\n",
       "      <td>944574</td>\n",
       "      <td>18843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1215</th>\n",
       "      <td>ENST00000327044.7</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>959256</td>\n",
       "      <td>15053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1259</th>\n",
       "      <td>ENST00000338591.8</td>\n",
       "      <td>ENSG00000187961.15</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>960584</td>\n",
       "      <td>965719</td>\n",
       "      <td>5135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1289</th>\n",
       "      <td>ENST00000379410.8</td>\n",
       "      <td>ENSG00000187583.11</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>966482</td>\n",
       "      <td>975865</td>\n",
       "      <td>9383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1326</th>\n",
       "      <td>ENST00000379407.7</td>\n",
       "      <td>ENSG00000187583.11</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>966502</td>\n",
       "      <td>975008</td>\n",
       "      <td>8506</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id             gene_id gene_name       gene_type seqname  \\\n",
       "164   ENST00000641515.2   ENSG00000186092.7     OR4F5  protein_coding    chr1   \n",
       "584   ENST00000426406.4   ENSG00000284733.2    OR4F29  protein_coding    chr1   \n",
       "816   ENST00000332831.5   ENSG00000284662.2    OR4F16  protein_coding    chr1   \n",
       "1115  ENST00000616016.5  ENSG00000187634.13    SAMD11  protein_coding    chr1   \n",
       "1148  ENST00000618323.5  ENSG00000187634.13    SAMD11  protein_coding    chr1   \n",
       "1181  ENST00000342066.8  ENSG00000187634.13    SAMD11  protein_coding    chr1   \n",
       "1215  ENST00000327044.7  ENSG00000188976.11     NOC2L  protein_coding    chr1   \n",
       "1259  ENST00000338591.8  ENSG00000187961.15    KLHL17  protein_coding    chr1   \n",
       "1289  ENST00000379410.8  ENSG00000187583.11   PLEKHN1  protein_coding    chr1   \n",
       "1326  ENST00000379407.7  ENSG00000187583.11   PLEKHN1  protein_coding    chr1   \n",
       "\n",
       "       start     end  length  \n",
       "164    65419   71585    6166  \n",
       "584   450740  451678     938  \n",
       "816   685716  686654     938  \n",
       "1115  923923  944574   20651  \n",
       "1148  923923  944574   20651  \n",
       "1181  925731  944574   18843  \n",
       "1215  944203  959256   15053  \n",
       "1259  960584  965719    5135  \n",
       "1289  966482  975865    9383  \n",
       "1326  966502  975008    8506  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genecode_sub.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2183268</th>\n",
       "      <td>ENST00000612152.4</td>\n",
       "      <td>ENSG00000000003.16</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100627109</td>\n",
       "      <td>100637104</td>\n",
       "      <td>9995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2183247</th>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>ENSG00000000003.16</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100636806</td>\n",
       "      <td>9698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2183227</th>\n",
       "      <td>ENST00000373031.5</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100584936</td>\n",
       "      <td>100599885</td>\n",
       "      <td>14949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046719</th>\n",
       "      <td>ENST00000683466.1</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr20</td>\n",
       "      <td>50935020</td>\n",
       "      <td>50959140</td>\n",
       "      <td>24120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2046646</th>\n",
       "      <td>ENST00000371582.8</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr20</td>\n",
       "      <td>50934867</td>\n",
       "      <td>50958555</td>\n",
       "      <td>23688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666650</th>\n",
       "      <td>ENST00000509903.5</td>\n",
       "      <td>ENSG00000310517.1</td>\n",
       "      <td>CAST</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr5</td>\n",
       "      <td>96702806</td>\n",
       "      <td>96771679</td>\n",
       "      <td>68873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667153</th>\n",
       "      <td>ENST00000325674.11</td>\n",
       "      <td>ENSG00000310517.1</td>\n",
       "      <td>CAST</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr5</td>\n",
       "      <td>96743627</td>\n",
       "      <td>96774679</td>\n",
       "      <td>31052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667073</th>\n",
       "      <td>ENST00000508579.5</td>\n",
       "      <td>ENSG00000310517.1</td>\n",
       "      <td>CAST</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr5</td>\n",
       "      <td>96741874</td>\n",
       "      <td>96772813</td>\n",
       "      <td>30939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667114</th>\n",
       "      <td>ENST00000515663.5</td>\n",
       "      <td>ENSG00000310517.1</td>\n",
       "      <td>CAST</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr5</td>\n",
       "      <td>96743540</td>\n",
       "      <td>96772706</td>\n",
       "      <td>29166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1711171</th>\n",
       "      <td>ENST00000562153.6</td>\n",
       "      <td>ENSG00000310525.1</td>\n",
       "      <td>ENSG00000310525</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr16</td>\n",
       "      <td>72038099</td>\n",
       "      <td>72093557</td>\n",
       "      <td>55458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65049 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              transcript_id             gene_id        gene_name  \\\n",
       "2183268   ENST00000612152.4  ENSG00000000003.16           TSPAN6   \n",
       "2183247   ENST00000373020.9  ENSG00000000003.16           TSPAN6   \n",
       "2183227   ENST00000373031.5   ENSG00000000005.6             TNMD   \n",
       "2046719   ENST00000683466.1  ENSG00000000419.14             DPM1   \n",
       "2046646   ENST00000371582.8  ENSG00000000419.14             DPM1   \n",
       "...                     ...                 ...              ...   \n",
       "666650    ENST00000509903.5   ENSG00000310517.1             CAST   \n",
       "667153   ENST00000325674.11   ENSG00000310517.1             CAST   \n",
       "667073    ENST00000508579.5   ENSG00000310517.1             CAST   \n",
       "667114    ENST00000515663.5   ENSG00000310517.1             CAST   \n",
       "1711171   ENST00000562153.6   ENSG00000310525.1  ENSG00000310525   \n",
       "\n",
       "              gene_type seqname      start        end  length  \n",
       "2183268  protein_coding    chrX  100627109  100637104    9995  \n",
       "2183247  protein_coding    chrX  100627108  100636806    9698  \n",
       "2183227  protein_coding    chrX  100584936  100599885   14949  \n",
       "2046719  protein_coding   chr20   50935020   50959140   24120  \n",
       "2046646  protein_coding   chr20   50934867   50958555   23688  \n",
       "...                 ...     ...        ...        ...     ...  \n",
       "666650   protein_coding    chr5   96702806   96771679   68873  \n",
       "667153   protein_coding    chr5   96743627   96774679   31052  \n",
       "667073   protein_coding    chr5   96741874   96772813   30939  \n",
       "667114   protein_coding    chr5   96743540   96772706   29166  \n",
       "1711171  protein_coding   chr16   72038099   72093557   55458  \n",
       "\n",
       "[65049 rows x 8 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " genecode_sub.sort_values(by=[\"gene_id\", \"length\"], ascending=[True, False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select longest transcript for each gene\n",
    "genecode_sub = genecode_sub.sort_values(by=[\"gene_id\", \"length\"], ascending=[True, False])\n",
    "genecode_sub = genecode_sub.drop_duplicates(subset=[\"gene_id\"], keep=\"first\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sort by chrom, start, end\n",
    "genecode_sub = genecode_sub.sort_values(by=[\"seqname\", \"start\", \"end\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>6166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>923923</td>\n",
       "      <td>944574</td>\n",
       "      <td>20651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1215</th>\n",
       "      <td>ENST00000327044.7</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>959256</td>\n",
       "      <td>15053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2225533</th>\n",
       "      <td>ENST00000361335.1</td>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrM</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2225540</th>\n",
       "      <td>ENST00000361381.2</td>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrM</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>1377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2225554</th>\n",
       "      <td>ENST00000361567.2</td>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrM</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>1811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2225560</th>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2225568</th>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>1140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20092 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             transcript_id             gene_id gene_name       gene_type  \\\n",
       "164      ENST00000641515.2   ENSG00000186092.7     OR4F5  protein_coding   \n",
       "584      ENST00000426406.4   ENSG00000284733.2    OR4F29  protein_coding   \n",
       "816      ENST00000332831.5   ENSG00000284662.2    OR4F16  protein_coding   \n",
       "1115     ENST00000616016.5  ENSG00000187634.13    SAMD11  protein_coding   \n",
       "1215     ENST00000327044.7  ENSG00000188976.11     NOC2L  protein_coding   \n",
       "...                    ...                 ...       ...             ...   \n",
       "2225533  ENST00000361335.1   ENSG00000212907.2   MT-ND4L  protein_coding   \n",
       "2225540  ENST00000361381.2   ENSG00000198886.2    MT-ND4  protein_coding   \n",
       "2225554  ENST00000361567.2   ENSG00000198786.2    MT-ND5  protein_coding   \n",
       "2225560  ENST00000361681.2   ENSG00000198695.2    MT-ND6  protein_coding   \n",
       "2225568  ENST00000361789.2   ENSG00000198727.2    MT-CYB  protein_coding   \n",
       "\n",
       "        seqname   start     end  length  \n",
       "164        chr1   65419   71585    6166  \n",
       "584        chr1  450740  451678     938  \n",
       "816        chr1  685716  686654     938  \n",
       "1115       chr1  923923  944574   20651  \n",
       "1215       chr1  944203  959256   15053  \n",
       "...         ...     ...     ...     ...  \n",
       "2225533    chrM   10470   10766     296  \n",
       "2225540    chrM   10760   12137    1377  \n",
       "2225554    chrM   12337   14148    1811  \n",
       "2225560    chrM   14149   14673     524  \n",
       "2225568    chrM   14747   15887    1140  \n",
       "\n",
       "[20092 rows x 8 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genecode_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecode_sub[['gene_name','gene_id', 'transcript_id']].to_csv(\"data/gencode/v47/gencode.v47.basic.annotation.parsed.pc.longest_tx.tsv\", sep=\"\\t\", index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"rdhs-activity.py\"\"\"\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Snakemake\n",
    "CCRES = snakemake.input[0]  # type: ignore\n",
    "SIGNAL_SUMMARY = snakemake.input[1]  # type: ignore\n",
    "OUTPUT = snakemake.output[0]  # type: ignore\n",
    "\n",
    "\n",
    "# ------------- #\n",
    "# Functions     #\n",
    "# ------------- #\n",
    "\n",
    "\n",
    "def read_rdhslist(filepath: str) -> pd.DataFrame:\n",
    "    \"\"\"Returns CCRE bed file as DataFrame\"\"\"\n",
    "    return pd.read_csv(\n",
    "        filepath,\n",
    "        sep=\"\\t\",\n",
    "        header=None,\n",
    "        engine=\"c\",\n",
    "        usecols=[3],\n",
    "        names=[\"rDHS\"],\n",
    "        dtype={\"rDHS\": str},\n",
    "    )[\"rDHS\"].to_list()\n",
    "\n",
    "\n",
    "def read_signal_summary(filepath: str) -> pd.DataFrame:\n",
    "    dtype = {\n",
    "        \"rDHS\": str,\n",
    "        \"sum_signal\": float,\n",
    "        \"num_signal\": int,\n",
    "    }\n",
    "    return pd.read_csv(\n",
    "        filepath,\n",
    "        sep=\"\\t\",\n",
    "        dtype=dtype,\n",
    "        engine=\"c\",\n",
    "    )\n",
    "\n",
    "\n",
    "def main():\n",
    "    \"\"\"Main\"\"\"\n",
    "    # Read inputs\n",
    "    rdhs_list = read_rdhslist(CCRES)\n",
    "    signal_summary = read_signal_summary(SIGNAL_SUMMARY)\n",
    "\n",
    "    # Subset to rDHS to known ccre\n",
    "    signal_summary = signal_summary[signal_summary[\"rDHS\"].isin(rdhs_list)]\n",
    "\n",
    "    # Calculate activity\n",
    "    signal_summary[\"activity\"] = signal_summary[\"sum_signal\"] / np.sqrt(\n",
    "        signal_summary[\"num_signal\"]\n",
    "    )\n",
    "\n",
    "    # Add quantile across all rDHS\n",
    "    signal_summary[\"quantile_rdhswide\"] = pd.qcut(\n",
    "        signal_summary[\"activity\"], 100, labels=[i for i in range(1, 101)]\n",
    "    )\n",
    "\n",
    "    # Tidy up\n",
    "    signal_summary = signal_summary[[\"rDHS\", \"activity\", \"quantile_rdhswide\"]].round(4)\n",
    "    dtypes = {\"rDHS\": str, \"activity\": float, \"quantile_rdhswide\": int}\n",
    "    signal_summary = signal_summary.astype(dtypes)\n",
    "\n",
    "    # Write output\n",
    "    signal_summary.to_csv(OUTPUT, sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "active",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
